jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java

```diff
@@ -114,64 +117,89 @@ public void handle(String target, Request baseRequest, HttpServletRequest reques
         }
 
         response.setStatus(HttpServletResponse.SC_NOT_FOUND);
-        response.setContentType(MimeTypes.Type.TEXT_HTML.toString());
+        response.setContentType(MimeTypes.Type.TEXT_HTML_UTF_8.toString());
 
-        try (ByteArrayISO8859Writer writer = new ByteArrayISO8859Writer(1500);)
+        try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+             OutputStreamWriter writer = new OutputStreamWriter(outputStream, UTF_8))
         {
-            writer.write("<HTML>\n<HEAD>\n<TITLE>Error 404 - Not Found");
-            writer.write("</TITLE>\n<BODY>\n<H2>Error 404 - Not Found.</H2>\n");
-            writer.write("No context on this server matched or handled this request.<BR>");
-            writer.write("Contexts known to this server are: <ul>");
+            writer.append("<!DOCTYPE html>\n");
+            writer.append("<html lang=\"en\">\n<head>\n");
+            writer.append("<title>Error 404 - Not Found</title>\n");
+            writer.append("<meta charset=\"utf-8\">\n");
+            writer.append("<style>body { font-family: sans-serif; } table, td { border: 1px solid #333; } td, th { padding: 5px; } thead, tfoot { background-color: #333; color: #fff; } </style>\n");
+            writer.append("</head>\n<body>\n");
+            writer.append("<h2>Error 404 - Not Found.</h2>\n");
+            writer.append("<p>No context on this server matched or handled this request.</p>\n");
+            writer.append("<p>Contexts known to this server are:</p>\n");
 
             Server server = getServer();
             Handler[] handlers = server==null?null:server.getChildHandlersByClass(ContextHandler.class);
 
+            writer.append("<table class=\"contexts\"><thead><tr>");
+            writer.append("<th>Context Path</th>");
+            writer.append("<th>Display Name</th>");
+            writer.append("<th>State</th>");
+            writer.append("<th>LifeCycle</th>");
+            writer.append("</tr></thead><tbody>\n");
+
             for (int i=0;handlers!=null && i<handlers.length;i++)
             {
+                writer.append("<tr><td>");
+                // Context Path
                 ContextHandler context = (ContextHandler)handlers[i];
+
+                StringBuilder href = new StringBuilder();
+
+                String contextPath = context.getContextPath();
+                href.append(contextPath);
```

Comment:
Should the `contextPath` be escaped?

Reply:
Escaping here would be interesting ...
For it to work, the user would have to use the URL escaped version, aka `ContextPath.setContextPath("/aaa%2Fbbb");`
So we wouldn't need to escape it here.